W-16141419: Adding Proxy configuration page for mule gw (#610)
mulesoft/docs-gateway: mule-gateway/modules/ROOT/pages/mule-gateway-proxy-config.adoc

= Route Mule Gateway Traffic Through a Proxy Server
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> You can configure the Mule Gateway to route traffic through the proxy server to API Manager if your Mule Runtime engine runs inside a firewall that restricts external communication through a proxy server.

Reviewer: ```suggestion
You can configure Mule Gateway to route traffic through the proxy server to API Manager if your Mule runtime engine runs inside a firewall that restricts external communication through a proxy server.
```

Author: Ok

Mule Gateway supports Basic, NTLM, and Kerberos v5 (via SPNEGO) authentication schemes for proxied communication.

[[configure-the-proxy-server]]
== Configure the Proxy Server

> To configure Mule Gateway to connect through a proxy server, you must set up your proxy server configuration in a `$MULE_HOME/conf/wrapper.conf` file. Add your proxy server information to the following properties:

Reviewer: ```suggestion
To configure Mule Gateway to connect through a proxy server, set up your proxy server configuration by adding the following properties in a `$MULE_HOME/conf/wrapper.conf` file:
```
Optional suggestion to make this shorter. Please check if it is accurate.

Author: Yep, it's ok

[source,text,linenums]
----
[libdefaults]
    kdc_realm = service.ws.apache.org
    default_realm = service.ws.apache.org
    udp_preference_limit = 1
    kdc_tcp_port = 64718

[realms]
    service.ws.apache.org = {
        kdc = localhost:64718
    }
----

==== Kerberos JAAS Path
The property `anypoint.platform.proxy_kerberos_jaas_path` specifies the path to the JAAS file for authentication. An example of this file is:

[source,text,linenums]
----
alice {
    com.sun.security.auth.module.Krb5LoginModule required
    refreshKrb5Config=true
    useKeyTab=false
    keyTab="/my/krbwd/alice.keytab"
    principal="alice";
};
com.sun.security.jgss.login {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
com.sun.security.jgss.initiate {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
com.sun.security.jgss.accept {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
----

IMPORTANT:

> Set the `useKeyTab` to `false` for the principal being authenticated (in this example, `alice`), so that it uses the credentials specified for `anypoint.platform.proxy_username` and `anypoint.platform.proxy_password`.

Reviewer: ```suggestion
Set `useKeyTab` to `false` for the principal being authenticated (in this example, `alice`), so that Mule Gateway uses the credentials specified for `anypoint.platform.proxy_username` and `anypoint.platform.proxy_password`.
```
The meaning of "it" was a bit unclear, please confirm it if is ok to use "Kerberos" in this case.

Author: It would actually be the Mule Gateway using Kerberos as authentication method. Not Kerberos itself, maybe we can just say Mule Gateway

> = Connect the Mule Gateway Through a Proxy Server

Reviewer 2: ```suggestion
= Route Mule Gateway Traffic Through a Proxy Server
```

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> If your Mule runtime engine runs inside a firewall that restricts external communication through a proxy server, you can configure the Mule Gateway to route traffic through the proxy server to API Manager.

Reviewer 3: ```suggestion
You can configure the Mule Gateway to route traffic through the proxy server to API Manager if your Mule Runtime engine runs inside a firewall that restricts external communication through a proxy server.
```

> The Mule Gateway supports the Basic, NTLM and Kerberos v5 (via SPNEGO) authentication schemes for proxied communication.

Reviewer 3: ```suggestion
Mule Gateway supports Basic, NTLM, and Kerberos v5 (via SPNEGO) authentication schemes for proxied communication.
```

> == Configure the Mule Gateway to Connect Through a Proxy Server

Reviewer 2: ```suggestion
== Configure the Proxy Server
```

> To configure Mule Gateway to connect through a proxy server, you must set up your proxy server configuration in a `$MULE_HOME/conf/wrapper.conf` file:

Reviewer 2: remove

Author: Why? Although it might sound obvious, Kerberos requires extra config files, so making it clear that these properties are for the wrapper.conf seems ok to me

> To configure the Mule Gateway to connect through a proxy server, you must set up your proxy server configuration in the `$MULE_HOME/conf/wrapper.conf` file.

Reviewer 3: ```suggestion
To configure Mule Gateway to connect through a proxy server, you must set up your proxy server configuration in a `$MULE_HOME/conf/wrapper.conf` file:
```

> === Set Up Proxy Server Configuration in the wrapper.conf File

Reviewer 3: ```suggestion
```

> To specify proxy server configuration in the `$MULE_HOME/conf/wrapper.conf` file,

Reviewer 3: ```suggestion
To configure the proxy server in the `$MULE_HOME/conf/wrapper.conf` file, add your proxy server information to the following properties:
```

> add your proxy server information to the following properties:

Reviewer 3: ```suggestion
```

* `wrapper.java.additional.<n>=-Danypoint.platform.proxy_port={port}`
* `wrapper.java.additional.<n>=-Danypoint.platform.proxy_username={username}`
* `wrapper.java.additional.<n>=-Danypoint.platform.proxy_password={password}`

> [IMPORTANT]

Reviewer 2: I do not believe this will render correctly

Author: It's rendering like this published doc: https://docs.mulesoft.com/runtime-manager/rtm-agent-proxy-config#set-up-proxy-server-configuration-in-the-wrapper-conf-file

> These parameters are additional options to pass to Java during launch. The <n> element refers to the number of additional parameters in the configuration, starting from `1` and incrementing sequentially without any gaps. 

Reviewer 2: ```suggestion
The parameters are additional options passed to Java during launch. The `<n>` element refers to the number of additional parameters in the configuration. Start from `1` and increment sequentially without skipping numbers. 
```

[IMPORTANT]

> These are additional parameters to pass to Java when it is launched. The <n> element refers to the number of additional parameters in the configuration. It is indicated with an integer number counting up from `1` and must follow a sequence without any gaps. 

Reviewer 3: ```suggestion
These parameters are additional options to pass to Java during launch. The <n> element refers to the number of additional parameters in the configuration, starting from `1` and incrementing sequentially without any gaps. 
```

> With the provided credentials, Mule Gateway will prioritize Basic Authentication over NTLM. If Basic Authentication fails, the Gateway will attempt to authenticate with NTLM. +

Reviewer 2: ```suggestion
With the provided credentials, Mule Gateway prioritizes Basic Authentication over NTLM. If Basic Authentication fails, Mule Gateway attempts to authenticate with NTLM.

```

> For Kerberos authentication, proceed to the next section.

Reviewer 2: remove

Author: Remove the `For Kerberos authentication, proceed to the next section.`?

> With the credentials provided, Mule Gateway will prefer Basic Authentication over NTLM. Failure to authenticate via Basic Auth would make Gateway attempt login with NTLM. For Kerberos, extra properties are required that are explained in the next section.

Reviewer 3: ```suggestion
With the provided credentials, Mule Gateway will prioritize Basic Authentication over NTLM. If Basic Authentication fails, the Gateway will attempt to authenticate with NTLM. +
For Kerberos authentication, proceed to the next section.
```

=== Using Kerberos for Proxy Authentication

> Alongside the properties described in the previous section, add the following properties:

Reviewer 3: ```suggestion
To authenticate using Kerberos, you need to add all the properties described in the previous section, plus the following new properties:

* `wrapper.java.additional.<n>=-Danypoint.platform.proxy_kerberos_krb5_path={/path/to/krb5.conf}`
* `wrapper.java.additional.<n>=-Danypoint.platform.proxy_kerberos_jaas_path={/path/to/kerberos.jaas}`

> NOTE: Kerberos authentication is only supported in Mule runtime versions 4.3.x and 4.4.x released after July 2, 2024 and Mule runtime versions 4.6.5, 4.7.1 and later.

Reviewer 3: ```suggestion
NOTE: Kerberos authentication is only supported in Mule runtime versions 4.3.x and 4.4.x, and Mule runtime versions 4.6.5, 4.7.1 and later.
```

Author: In this we should keep that `released after July 2, 2024`. 4.3.x and 4.4.x mule distributions have a timestamp that distinguish when the patch was released. They don't follow the patch increment and 4.6.x and 4.7.x that go onto 4.6.5 and 4.7.1 respectively. So it's important to keep the July 2 for customers to know as from which 4.3.x/4.4.x release the feature is available

==== Kerberos KRB5 Path

> Representing property `anypoint.platform.proxy_kerberos_krb5_path`, the value entered here is the path to the `krb5.conf` file used to locate the KRB5 server. An example of such file is:

Reviewer 3: ```suggestion
The property `anypoint.platform.proxy_kerberos_krb5_path` specifies the path to the `krb5.conf` file used to locate the KRB5 server. An example of this file is:
```

[source,text,linenums]
----
[libdefaults]
    kdc_realm = service.ws.apache.org
    default_realm = service.ws.apache.org
    udp_preference_limit = 1
    kdc_tcp_port = 64718

[realms]
    service.ws.apache.org = {
        kdc = localhost:64718
    }
----

==== Kerberos JAAS Path

> Representing property `anypoint.platform.proxy_kerberos_jaas_path`, the value entered here is the path to the JAAS file for authentication. An example of such file is:

Reviewer 3: ```suggestion
The property `anypoint.platform.proxy_kerberos_jaas_path` specifies the path to the JAAS file for authentication. An example of this file is:
```

[source,text,linenums]
----
alice {
    com.sun.security.auth.module.Krb5LoginModule required
    refreshKrb5Config=true
    useKeyTab=false
    keyTab="/my/krbwd/alice.keytab"
    principal="alice";
};
com.sun.security.jgss.login {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
com.sun.security.jgss.initiate {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
com.sun.security.jgss.accept {
  com.sun.security.auth.module.Krb5LoginModule required
  useKeyTab=true
  storeKey=true
  keyTab="/my/krbwd/alice.keytab"
  principal="alice";
};
----

> It's important to set the `useKeyTab` to false for the principal being authenticated, in this case `alice` so that it takes the credentials set up for `anypoint.platform.proxy_username` and `anypoint.platform.proxy_password`.

Reviewer 3: ```suggestion
IMPORTANT:
Set the `useKeyTab` to `false` for the principal being authenticated (in this example, `alice`), so that it uses the credentials specified for `anypoint.platform.proxy_username` and `anypoint.platform.proxy_password`.
```
